core/vm/evm.go

```diff
@@ -477,5 +543,49 @@ func (evm *EVM) Create2(caller ContractRef, code []byte, gas uint64, endowment *
 	return evm.create(caller, codeAndHash, gas, endowment, contractAddr)
 }
 
+func (evm *EVM) CreateETX(toAddr common.Address, fromAddr common.Address, etxGasLimit uint64, etxGasPrice *big.Int, etxGasTip *big.Int, etxData []byte, etxAccessList types.AccessList, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error) {
+
+	// Verify address is not in context
+	if toAddr.IsInChainScope() {
+		return []byte{}, 0, fmt.Errorf("%x is in chain scope, but CreateETX was called\n", toAddr)
+	}
+	if gas < params.ETXGas {
+		return []byte{}, 0, fmt.Errorf("CreateETX error: %d is not sufficient gas, required amount: %d", gas, params.ETXGas)
+	}
+
+	fee := big.NewInt(0)
+	fee.Add(etxGasTip, etxGasPrice)
+	fee.Mul(fee, big.NewInt(int64(etxGasLimit)))
+	total := big.NewInt(0)
+	total.Add(value, fee)
+	// Fail if we're trying to transfer more than the available balance
+	if total.Sign() == 0 || !evm.Context.CanTransfer(evm.StateDB, fromAddr, total) {
+		return []byte{}, 0, fmt.Errorf("CreateETX: %x cannot transfer %d\n", fromAddr, total.Uint64())
+	}
+	if err := evm.StateDB.SubBalance(fromAddr, total); err != nil {
+		return []byte{}, 0, fmt.Errorf("%x CreateETX error: %s\n", fromAddr, err.Error())
+	}
+
+	globalNonce, err := evm.StateDB.GetNonce(common.ZeroAddr)
+	if err != nil {
+		return []byte{}, 0, fmt.Errorf("%x CreateETX error: %s\n", fromAddr, err.Error())
+	}
+
+	// create external transaction
+	etxInner := types.ExternalTx{Value: value, To: &toAddr, Sender: fromAddr, GasTipCap: etxGasTip, GasFeeCap: etxGasPrice, Gas: etxGasLimit, Data: etxData, AccessList: etxAccessList, Nonce: globalNonce}
+	etx := types.NewTx(&etxInner)
+
+	evm.ETXCacheLock.Lock()
+	evm.ETXCache = append(evm.ETXCache, etx)
```

Comment:
@jdowning100 wrote this code, so he is more familiar, but even if the EVM engine is locked, it doesn't hurt to lock this too... it has almost zero cost in that case. However, if the EVM is not locked, then failing to lock this could expose a painful bug. Why not just lock it?

@jdowning100 please feel free to chime in

Reply:
I don't believe it is locked. In addition we want to support parallel execution which will require locking. However if you find where the EVM is actually locked per execution I'm willing to remove the lock for now.

Reply:
I saw the same lock used in some other place. So this lock is necessary.